src/curry/curry.js

```diff
@@ -32,4 +31,33 @@ const curry: CurryType = <A>(fn) => (...args) =>
     ? fn(...args)
     : (...rest) => curry(fn)(...args, ...rest)
 
-export { curry }
+/**
+ * Convert a curried function of arity n to a simple function with n parameters.
+ *
+ * @param fn
+ * The function to apply
+ *
+ * @param args
+ * The arguments to apply, in order
+ *
+ * @return
+ * The result of calling the function on each argument in turn,
+ * until a non-function is the return value
+ *
+ * @example
+ * uncurry(a => b => c => a + b * c)(1, 2, 3) = 7
+ */
+const uncurry: UncurryType = fn => (arg, ...args) => {
+  // The one liner version of this function doesn't type check because
+  // the type of of `fn(arg)` could change between each invocation. `fn`
+  // is not known to be pure because we can't say anything about its type
+  // in advance - in fact, we don't even know its arity.
+  //
+  // If we knew how to type `fn`, we could write:
+  // typeof fn(arg) === "function" ? uncurry(fn(arg))(...args) : fn(arg)
+  const result = fn(arg)
+
+  typeof result === "function" ? uncurry(result)(...args) : result
```

Comment:
This should be changed so that the partially applied function is returned if `args` runs out before we find a non-function result.